Source/Core/Common/BitUtils.h

```diff
@@ -33,78 +36,180 @@ constexpr size_t BitSize() noexcept
-/// @param  bit The bit to extract.
-///
-/// @tparam T   The type of the value.
+/// @param  start  The bit to extract. Bit 0 is the least-significant bit.
+/// @param  host   The host value to extract the bit from. Must satisfy std::integral.
 ///
 /// @return The extracted bit.
 ///
-template <typename T>
-constexpr T ExtractBit(const T src, const size_t bit) noexcept
+template <std::integral T>
+constexpr bool ExtractBit(const size_t start, const T host) noexcept
 {
-  return (src >> bit) & static_cast<T>(1);
+  DEBUG_ASSERT(start < BitSize<T>());  // Bit out of range
+  return static_cast<bool>((host >> start) & T{1});
 }
 
 ///
 /// Extracts a bit from a value.
 ///
-/// @param  src The value to extract a bit from.
-///
-/// @tparam bit The bit to extract.
-/// @tparam T   The type of the value.
+/// @tparam start  The bit to extract. Bit 0 is the least-significant bit.
+/// @param  host   The host value to extract the bit from. Must satisfy std::integral.
 ///
 /// @return The extracted bit.
 ///
-template <size_t bit, typename T>
-constexpr T ExtractBit(const T src) noexcept
+template <size_t start, std::integral T>
+constexpr bool ExtractBit(const T host) noexcept
 {
-  static_assert(bit < BitSize<T>(), "Specified bit must be within T's bit width.");
-
-  return ExtractBit(src, bit);
+  static_assert(start < BitSize<T>(), "Bit out of range");
+  return ExtractBit(start, host);
 }
 
+// TODO C++23: The x86 CPU extension "Bit Manipulation Instruction Set 1" provides unsigned bit
+// field extraction instructions which could theoretically be faster than the idiomatic way of doing
+// this. However, the only way to create specializations using BEXTR intrinsics while remaining
+// constexpr requires consteval if statements. Even then, BMI1 is from 2012-ish (SSE2 is Dolphin
+// Emulator's only hard-required x86 CPU extension atm), so these specializations would only really
+// be available to end-users using GCC/Clang who enable -march=native when compiling.
+// Bit Manipulation Instruction Set 2's PEXT/PDEP instructions also are worth looking at.
+
+///
+/// Extracts a zero-extended range of bits from a value.
 ///
-/// Extracts a range of bits from a value.
+/// @param  start  The beginning of the bit range. Bit 0 is the least-significant bit.
+/// @param  width  The width of the bit range in bits.
+/// @param  host   The host value to extract the bits from. Must satisfy std::integral.
+///
+/// @return The extracted bits.
+///
+template <std::integral T>
+constexpr std::make_unsigned_t<T> ExtractBitsU(const size_t start, const size_t width,
+                                               const T host) noexcept
+{
+  // TODO: would pre-calculating the mask would be faster than calculating it at run-time?
+  DEBUG_ASSERT(start + width <= BitSize<T>());  // Bitfield out of range
+  DEBUG_ASSERT(width > 0);                      // Bitfield is invalid (zero width)
+  const T mask = (T{1} << width) - T{1};
```

Comment:
FWIW one idea you could consider here is wrapping these constants in an `std::integral_constant`. Moving them to the type level like that results in a flatter AST, which makes it easier for optimizers to decide whether to inline functions these functions are used in.

This is something that probably shouldn't be done without measuring its practical impact first.